src/Agent.Sdk/SecretMasking/OssSecretMasker.cs

```diff
@@ -81,11 +106,132 @@ public void Dispose()
     {
         _secretMasker?.Dispose();
         _secretMasker = null;
+        _telemetry = null;
     }
 
     public string MaskSecrets(string input)
     {
-        return _secretMasker.MaskSecrets(input);
+        _telemetry?.ProcessInput(input);
+        return _secretMasker.MaskSecrets(input, _telemetry?.DetectionAction);
+    }
+
+    public void EnableTelemetry()
+    {
+        _telemetry ??= new Telemetry();
+    }
+
+    /// <summary>
+    /// If enabled via <see cref="TelemetryEnabled"/> opt-in, publishes
+    /// telemetry via the provided callback that handles sending it over the
+    /// wire.
+    ///
+    /// Always publishes a 'SecretMasker' event with overall stats, and may
+    /// publish 'SecretMaskerDetections' events mapping C3ID (12 byte
+    /// non-reversible seeded hash) to pattern moniker ('name'.'id') for
+    /// rule-based patterns with high entropy if any such detections are found.
+    /// </summary>
+    public void PublishTelemetry(PublishSecretMaskerTelemetryAction publishAction)
+    {
+        if (_telemetry == null)
+        {
+            return;
+        }
+
+        Dictionary<string, string> detectionData = null;
+        int events = 0;
+
+        foreach (var pair in _telemetry.Detections)
```

Comment:
>Detections [](http://example.com/codeflow?start=40&length=10)

It hurts my poor tired brain that we are here accessing a concurrent dictionary in all confidence that no one else is still partying on it.

If there's some requirement here in this method that threading must now be constrained to a single caller, that's unfortunate for a newly added interface method, arguably. In another model, what we'd do is just publish the Telemetry object. The masker consumer would know the current state of object access and retrieve the detections data.

Just thinking aloud here, want to hear what you say.

Reply:
> In another model, what we'd do is just publish the Telemetry object. The masker consumer would know the current state of object access and retrieve the detections data.

I haven't understood this idea yet.

Reply:
So to be unclear. In this other design, instead of calling EnableTelemetry, then Publish, you'd just push a Telemetry object onto the masker. When non-null, the masker would interact with the Telemetry object. The secret masker allocater would retrieve the telemetry object when appropriate to do stuff with the data (e.g., actually publish it).

Reply:
This is fixed now. Changing the callback design like that doesn't actually solve the issue and in fact making it more general would opens up more threading cases to think about IMHO.

But we now ensure that Telemetry cannot be published in the middle of an operation that produces telemetry. This removes the ability to publish more than once, but that was never needed.